### PR TITLE
MBT mass matrix and bias term computation methods

### PR DESCRIPTION
**Pull request by amcastro-tri**
This PR introduces `MultibodyTree::CalcMassMatrixViaInverseDynamics()` and `MultibodyTree<T>::CalcBiasTerm()`. These are unit tested in `double_pendulum_test.cc`.

Notice I also pushed here a bug fix in `body_node.h` where the wrong (the inverse) transformation was used.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/7053)

### CODE REVIEW

**Comment by amcastro-tri** [comment]
+@avalenzu for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-:-KuQ_-xtjQlrt9Fat2K6:bll7v0f)*

**Comment by avalenzu** [comment]
Review pass done. PTAL.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 277 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVjsjdMnXWoAYQk11u:-KuVjsjeIolkl93-W1si:bpqi485) ([raw file](https://github.com/robotlocomotion/drake/blob/e521017e389323425218496d5edf09b78bfcd67f/drake/multibody/multibody_tree/multibody_tree.cc#L277)):*
> ```cpp
>   DRAKE_DEMAND(static_cast<int>(F_BMo_W_array->size()) == get_num_bodies());
> 
>   DRAKE_DEMAND(tau_array != nullptr);
> ```

FYI (actually For My Information) Are `EigenPtr`s guaranteed to be non-null?

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 785 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVlTAhtvrIt7U75Nm8:-KuVlTAhtvrIt7U75Nm9:behtgy6) ([raw file](https://github.com/robotlocomotion/drake/blob/68ba7a41f5452bd9e8711fa4f9c9bb99876cdcfc/drake/multibody/multibody_tree/multibody_tree.h#L785)):*
> ```objc
>   ///   The context containing the state of the %MultibodyTree model. It stores
>   ///   the generalized positions q and the generalized velocities v.
>   /// @param[out] C
> ```

It seems confusing to call this `C` when the description above says that it's `C(q,v)v`.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 516 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVnIGEAS8ijY9PffVu:-KuVnIGEAS8ijY9PffVv:b-rjgs3a) ([raw file](https://github.com/robotlocomotion/drake/blob/68ba7a41f5452bd9e8711fa4f9c9bb99876cdcfc/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L516)):*
> ```cpp
>     model_->CalcBiasTerm(*context_, &C);
>     C_expected = acrobot_benchmark_.CalcCoriolisVector(
>             shoulder_angle, elbow_angle, shoulder_rate, elbow_rate);
> ```

It looks like `C` and `C_expected` are never actually compared? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-:-KuVnbf2ln9todMj9JGE:b837gv3)*

**Comment by amcastro-tri** [comment]
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 277 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVjsjdMnXWoAYQk11u:-KuZfCmWTJOuks8MUsiR:bvnir65) ([raw file](https://github.com/robotlocomotion/drake/blob/e521017e389323425218496d5edf09b78bfcd67f/drake/multibody/multibody_tree/multibody_tree.cc#L277)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

FYI (actually For My Information) Are `EigenPtr`s guaranteed to be non-null?
</blockquote></details>

BTW. First of all, in case you are not familiar with them yet, these are Drake types, i.e. `drake::EigenPtr`. You can find it in `eigen_types.h`, where you'll see the default constructor initializes it to `nullptr`.
These `EigenPtr` objects where introduced so that our Gstyle guide compliant code still looks like `MyMethod(input_vector, &output_vector)` when passing Eigen objects that are to be modified on output (notice the `&` makes it look like a pointer, as the GStyle guide requires).
So, bottom line, we now should be using `const Eigen::Ref<const VectorX>&` for Eigen types on input and `EigenPtr` for Eigen objects on output (though there still is code in master that keeps using non-const `Eigen::Ref`, hopefully to change soon).

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 785 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVlTAhtvrIt7U75Nm8:-KuZh7Go4MZpnXQ2_5OL:brjxe7) ([raw file](https://github.com/robotlocomotion/drake/blob/68ba7a41f5452bd9e8711fa4f9c9bb99876cdcfc/drake/multibody/multibody_tree/multibody_tree.h#L785)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

It seems confusing to call this `C` when the description above says that it's `C(q,v)v`.
</blockquote></details>

I refactored it to be `Cv` and updates doc accordingly.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 516 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVnIGEAS8ijY9PffVu:-KuZikzedJypu4HkEVdf:bjgjfsu) ([raw file](https://github.com/robotlocomotion/drake/blob/68ba7a41f5452bd9e8711fa4f9c9bb99876cdcfc/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L516)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

It looks like `C` and `C_expected` are never actually compared? 
</blockquote></details>

No wonder why these tests were passing! Good catch. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053)*

**Comment by avalenzu** [comment]
:lgtm: +@jwnimmer-tri  for platform review

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 277 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-KuVjsjdMnXWoAYQk11u:-Ku_1ktn2T4kQmb0uAmD:b-wsvasg) ([raw file](https://github.com/robotlocomotion/drake/blob/e521017e389323425218496d5edf09b78bfcd67f/drake/multibody/multibody_tree/multibody_tree.cc#L277)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW. First of all, in case you are not familiar with them yet, these are Drake types, i.e. `drake::EigenPtr`. You can find it in `eigen_types.h`, where you'll see the default constructor initializes it to `nullptr`.
These `EigenPtr` objects where introduced so that our Gstyle guide compliant code still looks like `MyMethod(input_vector, &output_vector)` when passing Eigen objects that are to be modified on output (notice the `&` makes it look like a pointer, as the GStyle guide requires).
So, bottom line, we now should be using `const Eigen::Ref<const VectorX>&` for Eigen types on input and `EigenPtr` for Eigen objects on output (though there still is code in master that keeps using non-const `Eigen::Ref`, hopefully to change soon).
</blockquote></details>

Thanks for the explanation!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-:-Ku_1zFhkB3E4HV8xI0D:brsjo7)*

**Comment by jwnimmer-tri** [comment]
:lgtm: with some FYIs.

---

Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 356 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_5yfCKUrgMCTABJJ_:-Ku_5yfDX9Acma74xS0C:b1ng2lr) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/multibody_tree.cc#L356)):*
> ```cpp
>   vc.InitializeToZero();
> 
>   // ======================================================================
> ```

OK Did they have a sale on `=` signs?  ;)  This seems like noise.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 792 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_5IZ5QleGvPMWA0s8:-Ku_5IZ5QleGvPMWA0s9:bcc3im2) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/multibody_tree.h#L792)):*
> ```objc
>   ///   proper size.
>   void CalcBiasTerm(
>       const systems::Context<T>& context, EigenPtr<VectorX<T>> Cv) const;
> ```

FYI You have line-wrapped this method and the prior one differently, even though they are nearly identical and could be done consistently.

Ditto in the cc file.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 937 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_5UovSqUvNeMlGZ8B:-Ku_5UovSqUvNeMlGZ8C:b-lm17x7) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/multibody_tree.h#L937)):*
> ```objc
>   //  - The position kinematics cache object is already updated to be in sync
>   //    with `context`.
>   //  - H is not the nullptr.
> ```

FYI `the nullptr` is odd; typical grammar would be just `nullptr`.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 553 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_7TxPkBRpeYojMM56:-Ku_7TxQOc5DUlFq_K8x:b-nkoc62) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L553)):*
> ```cpp
>         shoulder_angle, elbow_angle, shoulder_rate, elbow_rate);
>     EXPECT_TRUE(CompareMatrices(
>         C, C_expected, kTolerance, MatrixCompareType::relative));
> ```

FYI Borderline whether defective or not, but the above 4x copies are a bit repetitive.  Why not use a for loop over `{shoulder_rate, elbow_rate}`, for brevity?

Also FWIW It seems like many of the cases below could be googletest parameterized tests, which would place every set of constants into its own test case, have nice clear error messages, and be easy to add more combinations or permutations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-:-Ku_8N9OeQdAbfvb_myd:bviqlcg)*

**Comment by amcastro-tri** [comment]
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 356 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_5yfCKUrgMCTABJJ_:-Ku_AbNzE-A4jKxFme5q:b-4caf5y) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/multibody_tree.cc#L356)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK Did they have a sale on `=` signs?  ;)  This seems like noise.
</blockquote></details>

40% percent off man, irresistible! I'll invite you next time

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 792 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_5IZ5QleGvPMWA0s8:-Ku_BBPWFH1k4JHv8cP-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/multibody_tree.h#L792)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI You have line-wrapped this method and the prior one differently, even though they are nearly identical and could be done consistently.

Ditto in the cc file.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 937 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_5UovSqUvNeMlGZ8B:-Ku_Bo2NVbloxvUmdjDz:b-76ifnx) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/multibody_tree.h#L937)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `the nullptr` is odd; typical grammar would be just `nullptr`.
</blockquote></details>

Thanks. I've been making this mistake for a while it looks like. I changed in other places as well.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 553 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_7TxPkBRpeYojMM56:-Ku_Bx_bNIowx8TBmt1D:b-g5bvnt) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L553)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Borderline whether defective or not, but the above 4x copies are a bit repetitive.  Why not use a for loop over `{shoulder_rate, elbow_rate}`, for brevity?

Also FWIW It seems like many of the cases below could be googletest parameterized tests, which would place every set of constants into its own test case, have nice clear error messages, and be easy to add more combinations or permutations.
</blockquote></details>

I couldn't agree with you more @jwnimmer-tri. I actually have a parametrized version (though on other parameters) in my dev branch. I just need to find the man hours to push it! I'll pass this time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053)*

**Comment by jwnimmer-tri** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 553 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7053#-Ku_7TxPkBRpeYojMM56:-Ku_E08yQvCr_JyGpSle:bv0bt25) ([raw file](https://github.com/robotlocomotion/drake/blob/4dd788c0e3a428c2a0408c0a7cb52ea6614ba683/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L553)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I couldn't agree with you more @jwnimmer-tri. I actually have a parametrized version (though on other parameters) in my dev branch. I just need to find the man hours to push it! I'll pass this time.
</blockquote></details>

OK Awesome!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053)*

**Comment by avalenzu** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7053)*

**Comment by jwnimmer-tri** [comment]
FYI This is ready to merge, other than the curate.  Please remember to edit the squash message to be concise.
